### PR TITLE
Fix Issue 22552 - moveEmplace wipes context pointer of nested struct …

### PR DESCRIPTION
**Pull request by radcapricorn**
…contained in non-nested struct

### CODE REVIEW

**Comment by radcapricorn** [comment]
ping @kinke

**Comment by kinke** [comment]
Thanks for working on this. After a quick glance, I think it should try a bit harder to prevent the recursion, in case the source instance itself is the only holder of a context pointer.
```
void foo() {
    int i;
    struct S {
        int a0, a1, a2, a3, a4, a5, …;
        ~this() { ++i; }
    }
}
```
For the struct above, it looks like you'd currently reset all `int` fields individually to zero (via `*cast(ubyte[4]) &an = 0` for N fields) instead of memsetting once and leaving the context pointer untouched.

File: src/core/lifetime.d
Comment (RazvanN7) [inline src/core/lifetime.d:2165]: Hmm, I don't know exactly what the mechanics for context pointers and unions are, but since this code is working:

```d
auto fun()
{
    int a;
    struct S { int f() {return a;} }
    
    union T
    {
         S b;   
    }
    
    T t;
    return t;
}

void main()
{
    auto s = fun();
    int t = s.b.f();
}
```

Doesn't that mean that `T` has a context pointer?

Comment (radcapricorn) [reply src/core/lifetime.d:2165]: Unions can yield false positives. They're not discriminated. The danger of wiping context pointers is destructors, and unions don't have those. It's the caller's job to ascertain discrimination, including safe default discriminator in .init.

If ever needed, we could add a `mayHaveContextPointers` that would include unions and static arrays of void.

Comment (radcapricorn) [reply src/core/lifetime.d:2168]: Your code actually raises a question to the [spec](https://dlang.org/spec/struct.html#default_union_init):
> Unions are by default initialized to whatever the Initializer for the first field is, and if none is supplied, to the **default initializer** for the first field's type

You're not supplying initializer, so default one should be used. But default initializer would not have any context pointer set (i.e. `S.init.tupleof[$-1] is null` - which is the cause of this bug in the first place), yet in the above code it is being set. Rules for this particular case need to be specified explicitly.

**Comment by RazvanN7** [comment]
I will leave this for @kinke 's approval.

**Comment by radcapricorn** [comment]
> I will leave this for @kinke 's approval.

This needs more than just approval, this needs a review :) There are a couple things specifically mentioned in the comments, including what looks like another bug.

**Comment by RazvanN7** [comment]
What I meant is that I am not qualified to review to this :(

File: src/core/lifetime.d
Comment (kinke) [inline src/core/lifetime.d:2470]: It might make sense to try to combine both wipe-versions into one, as the basic static-if structure is the same - maybe as a variadic template, with a single optional `const scope ref T initializer` param.

File: src/core/lifetime.d
Comment (kinke) [inline src/core/lifetime.d:2488]: Thx for the `.offsetof` fix - the previous code assumed the context pointer to be right at the end, without any tail padding.

I haven't checked, but I assume some nested `align(64) struct S { int x; }` to contain the pointer at offset 8 [for a 64-bit target], which would leave the 48 bytes tail padding undefined. I guess the safe variant would be to save the ptr, memset/cpy the whole initializer, and then restore the ptr again. Otherwise equality/`is` identity of structs could be broken.

File: src/core/lifetime.d
Comment (kinke) [inline src/core/lifetime.d:2504]: I think the POD-ness check isn't required, is it?

File: src/core/lifetime.d
Comment (kinke) [inline src/core/lifetime.d:2218]: Good catch; I guess a dedicated PR with a little unittest of its own would be nicer.

File: src/core/lifetime.d
Comment (kinke) [reply src/core/lifetime.d:2488]: Ouch, and the recursive field-wise wiping is missing *all* paddings. :/

Comment (kinke) [reply src/core/lifetime.d:2488]: Maybe this could/should be simplified to:
1. blit src into dst
2. reset src to T.init, clearing all context pointers but initializing all gaps and paddings
3. if need be: recurse into fields and copy context pointers back from dst to src

Comment (kinke) [reply src/core/lifetime.d:2488]: Erm wait, this is the wiping of a supposedly valid instance - all paddings should be well-defined already. Please ignore. ;)

File: src/core/lifetime.d
Comment (radcapricorn) [reply src/core/lifetime.d:2470]: Is there a benefit to doing that? Going to be the same code, just harder to read, and the same amount of symbols generated. I mean, I can do that, just not sure if it's worth doing.

File: src/core/lifetime.d
Comment (radcapricorn) [reply src/core/lifetime.d:2488]: > some nested `align(64) struct S { int x; }`

Oh yeah, need to add a test for things with custom-aligned pointer, thanks!

File: src/core/lifetime.d
Comment (radcapricorn) [reply src/core/lifetime.d:2504]: It is, since the eventual else may do opAssign, and for reasons that are not known to me, presence of custom opAssign does not make a thing not POD :\ Mayhap that needs to be reported?

File: src/core/lifetime.d
Comment (radcapricorn) [reply src/core/lifetime.d:2218]: Aight, will do, thanks!

File: src/core/lifetime.d
Comment (kinke) [reply src/core/lifetime.d:2504]: Then moving it to the `hasElaborateAssign` etc. condition would do just fine too, right?

File: src/core/lifetime.d
Comment (kinke) [reply src/core/lifetime.d:2470]: > Cuz readers of such code definitely won't :D

On the contrary, my guess is that it would improve readability by reducing duplication and number of lines. After all, the only difference is doing a memset-to-0 or memcpy in 3 lines, and optionally passing the initializer for the 2 recursive calls.

Comment (radcapricorn) [reply src/core/lifetime.d:2470]: But the `Args...`, the constraint, the `static if (Args.length)`... Anyway, let's just do it and see how it reads.

File: src/core/lifetime.d
Comment (radcapricorn) [reply src/core/lifetime.d:2504]: Yeah. I think I added that condition some time after making the isPOD test and didn't look at that again.

**Comment by radcapricorn** [comment]
Assuming current implementation is satisfactory, the remaining questions are:
- whether to make the `hasContextPointers` public in `core.internal.traits`
- whether moving from structs that have head-const fields should be allowed. For that matter, the question is also whether `move` into anything const should be allowed.

**Comment by RazvanN7** [comment]
> whether to make the hasContextPointers public in core.internal.traits

cc @WalterBright @atilaneves 

> whether moving from structs that have head-const fields should be allowed. For that matter, the question is also whether move into anything const should be allowed.

I would say no (except maybe if the move happens in a constructor and the target is a member field and this is the first use of member inside the constructor - although this would require compiler support)

**Comment by atilaneves** [comment]
I would say no too but I defer to @WalterBright

File: src/core/lifetime.d
Comment (ljmf00) [inline src/core/lifetime.d:2181]: Since this import is a local import, can you be selective of what symbols to import?

File: src/core/lifetime.d
Comment (ljmf00) [inline src/core/lifetime.d:2506]: ditto

File: src/core/lifetime.d
Comment (ljmf00) [inline src/core/lifetime.d:2457]: This appears unused. If it is intentional, use it at the end and add a comment to avoid uncovered lines.

Comment (radcapricorn) [reply src/core/lifetime.d:2457]: Oh, good catch, thanks. Since my other fix for static arrays is now in this can now be moved to destructor.

**Comment by radcapricorn** [comment]
Are we merging this this century? ;)

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**
